content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+  of memory the cgroup must always retain. Hence, the container runtime did not
+  use the value of requested memory set in the Pod Spec. 
+
+  cgroup v2 introduced `memory.min` interface to specify the minimum amount of memory
+  a cgroup must always retain. If the memory usage of a cgroup is within its effective
+  min boundary, the cgroup’s memory won’t be reclaimed under any conditions. If 
+  there's no unprotected reclaimable memory available, OOM killer is invoked. Memory
+  QoS maps `memory.min` to `spec.containers[].resources.requests.memory` to ensure
+  the availability of memory for containers in Kubernetes Pods. 
+
+## How it works?
+
+### Cgroup v2 memory controller interfaces & Kubernetes container resources mapping
+
+Memory QoS uses the memory controller of cgroup v2 to guarantee memory resources in
+Kubernetes. cgroupv2 interfaces that this feature: 
+* `memory.max`
+* `memory.min`
+* `memory.high`.
+
+![](./memory-qos-cal.svg)
+
+`memory.max` is mapped to memory limit specified in the Pod spec. The kubelet (and 
+the container runtime) enforces the limit and prevent the container from using more
+than the configured resource limit. If a process in a container tries to consume
+more than the specified limit, kernel terminates the process with an out of
+memory Out of Memory (OOM) error. 
+
+
+![](./container-memory-max.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` is mapped to memory requests, which results in reservation of memory resources
+that can never reclaimed by the kernel. This is how Memory QoS ensures the availability of 
+memory for Kubernetes pods. If there's no unprotected reclaimable memory available, OOM 
+kill is invoked.
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+For memory protection, in addition to the original way of limiting memory usage, Memory QoS
+throttles workload approaching its memory limit, ensuring that the system is not overwhelmed
+by sporadic increases in memory usage. A new field, MemoryThrottlingFactor, is introduced in
+the KubeletConfiguration which is set to 0.9 by default. `memory.high` is calculated by using
+MemoryThrottlingFactor, `spec.containers[].resources.requests.memory` and
+`spec.containers[].resources.limits.memory` as in the formula below, and rounding down the
+value to the nearest page size:
+
+![](./container-memory-high.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+**Summary:**
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.max</td>
+        <td><code>memory.max</code> specifies the maximum memory limit, a container is allowed to use. If a container tries to consume more memory than the configured limit, kernel terminates the process with an Out of Memory (OOM) error. 
+        <br>
+        <br>
+        <i>We map it to the container's memory limit</i>
```

Comment:
This will work better as semantic HTML. For example:
```suggestion
        <td>
         <p><tt>memory.max</tt> specifies the maximum memory limit that a container is allowed to use.
            If a container tries to consume more memory than the configured limit, the kernel terminates
            the process with an Out of Memory (OOM) error.
        </p> 
        <p><em>Kubernetes maps it to the container's memory limit.</em></p>
```